python/lsst/rubintv/production/slac/utils.py

```diff
@@ -0,0 +1,162 @@
+    collection, which is non-standard. If/when the repo surgery is done to
+    chain this on to the standard calib collection, this util can be removed.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler.
+    instrument : `str`
+        The instrument name, either 'LSST-TS8' or 'LSSTCam'.
+
+    Returns
+    -------
+    camera : `lsst.afw.cameraGeom.Camera`
+        The camera object.
+    """
+    match instrument:
+        case 'LSSTCam':
+            return butler.get('camera', collections=['LSSTCam/calib/unbounded'], instrument='LSSTCam')
+        case 'LSST-TS8':
+            return butler.get('camera', collections=['LSST-TS8/calib/unbounded'], instrument='LSST-TS8')
+        case _:
+            raise ValueError("This utility function is just for getting LSST-TS8 and LSSTCam instruments,"
+                             f" got {instrument}")
+
+
+def fullAmpDictToPerCcdDicts(fullDict):
+    """Split a single, flat dict of per-amp values by detector.
+
+    This function exists because some eo_pipe plotting utils expect the data
+    in this format.
+
+    For example, a single dict like:
+    {'R12_S01_C05': 123,
+     'R12_S02_C06': 234,
+     ...
+    }
+
+    is turned into a dict of dicts, keyed by full detector name,
+    with the values in a dict of amp values:
+    {'R12_S01': {'C05': 123},
+     'R12_S02': {'C06': 234},
+     ...
+    }
+
+    Parameters
+    ----------
+    fullDict : `dict` [`str`, `float`]
+        The flat input dict of per-amp values, keyed like 'R12_S01_C05': 123.4
+
+    Returns
+    -------
+    perCcdDicts : `dict` [`str`, `dict` [`str`, `float`]]
+        The per-ccd dict of dicts, keyed by full detector name, with the values
+        keyed by amp name.
+    """
+    detectors = set(name.split('_C')[0] for name in fullDict.keys())
+    dicts = {k: {} for k in detectors}
+    for k, v in fullDict.items():
+        detName, ampName = k.split('_C')
+        dicts[detName]["C" + ampName] = v
+    return dicts
+
+
+def getDetectorsWithData(butler, expRecord, dataset):
+    """Get a list of detector ids for which the specified dataset exists.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        The butler.
+    expRecord : `lsst.daf.butler.DimensionRecord`
+        The exposureRecord containing the dataId.
+    dataset : `str`
+        The dataset type to find, e.g. 'raw'.
+
+    Returns
+    -------
+    detectorsWithData : `list` [`int`]
+        The list of detector numbers for which the dataset exists.
+    """
+    return sorted([det.id for det in butler.registry.queryDimensionRecords('detector',
+                                                                           dataId=expRecord.dataId,
+                                                                           datasets=dataset)])
+
+
+def waitForDataProduct(butler, expRecord, dataset, detector, timeout, cadence=1, logger=None):
```

Comment:
Could this be unified with other similar logic elsewhere, maybe in the `ButlerWatcher`?

Reply:
Uggg, you're right that that is _extremely_ similar to the `_waitForDataProduct` on the `BaseChannel`, this was part of the refactoring work. I think I will add a TODO to the code for now, and address that at the same time as making the new cleanroom code inherit from those base channels, as that all is part of the same coherent bit of work.